wrappers/core18.go

```diff
@@ -212,6 +267,86 @@ func writeSnapdServicesOnCore(s *snap.Info, inter interacter) error {
 	return nil
 }
 
+func undoSnapdServicesOnCore(s *snap.Info, sysd systemd.Systemd) error {
+	serviceUnits, err := filepath.Glob(filepath.Join(s.MountDir(), "lib/systemd/system/*.service"))
+	if err != nil {
+		return err
+	}
+	socketUnits, err := filepath.Glob(filepath.Join(s.MountDir(), "lib/systemd/system/*.socket"))
+	if err != nil {
+		return err
+	}
+	timerUnits, err := filepath.Glob(filepath.Join(s.MountDir(), "lib/systemd/system/*.timer"))
+	if err != nil {
+		return err
+	}
+	units := append(socketUnits, serviceUnits...)
+	units = append(units, timerUnits...)
+
+	for _, snapdUnit := range units {
+		unit := filepath.Base(snapdUnit)
+		coreUnit := filepath.Join(dirs.GlobalRootDir, "lib/systemd/system", unit)
+		writtenUnitPath := filepath.Join(dirs.SnapServicesDir, unit)
+		if !osutil.FileExists(writtenUnitPath) {
+			continue
+		}
+		existsInCore := osutil.FileExists(coreUnit)
+
+		if !existsInCore {
+			// new unit that did not exist on core, disable and stop
+			if err := sysd.Disable(unit); err != nil {
+				logger.Noticef("failed to disable %q: %v", unit, err)
+			}
+			if err := sysd.Stop(unit, snapdServiceStopTimeout); err != nil {
+				return err
+			}
+		}
+		if err := os.Remove(writtenUnitPath); err != nil {
+			return err
+		}
+		if !existsInCore {
+			// nothing more to do here
+			continue
+		}
+
+		isEnabled, err := sysd.IsEnabled(unit)
+		if err != nil {
+			return err
+		}
+		if !isEnabled {
+			if err := sysd.Enable(unit); err != nil {
+				return err
+			}
+		}
+
+		if unit == "snapd.socket" {
+			// do not start the socket, snap failover handler will
+			// restart it
+			continue
+		}
+		skipStart, err := snapdUnitSkipStart(coreUnit)
+		if err != nil {
+			return err
+		}
+		if !skipStart {
+			isActive, err := sysd.IsActive(unit)
+			if err != nil {
+				return err
+			}
+			if isActive {
+				if err := sysd.Restart(unit, snapdServiceStopTimeout); err != nil {
```

Comment:
Can we simply restart the unit unconditionally (without checking isActive)?

Reply:
Good point, it's likely unnecessary in the undo path. I'd be more confident if this ran through spread test though. I'll add a TODO here.